### PR TITLE
HPCC-25166 Turn off translation if alwaysECL is used globally.

### PR DESCRIPTION
**Pull request by jakesmith**
Using alwaysECL globally is highly likely to cause deserialization
issues and crashes. It should not be used.
Existing uses will cause a logged warning to be issued and
translation to be turned off.

Also warn if alwaysDisk is used globally.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [x] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [ ] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@richardkchapman - please review

File: rtl/eclrtl/rtldynfield.cpp [hunk 1]
```diff
@@ -41,17 +41,16 @@ extern ECLRTL_API RecordTranslationMode getTranslationMode(const char *val, bool
...
+            WARNLOG("alwaysDisk translation mode should only ever be used via a HINT");
         return RecordTranslationMode::AlwaysDisk;
     }
     else if (strieq(val, "alwaysECL") || strieq(val, "ecl"))
     {
         if (!isLocal)
-            throw makeStringException(0, "alwaysECL translation mode can only be set via HINT");
-        return RecordTranslationMode::AlwaysECL;
+            return RecordTranslationMode::AlwaysECL;
+        WARNLOG("alwaysECL translation mode should only ever be used via a HINT. Turning translation off.");
```
Comment (richardkchapman) [inline rtl/eclrtl/rtldynfield.cpp:51]: I would word differently - "Ignoring alwaysECL translation global setting - use with HINT to set locally"

**Review by richardkchapman — CHANGES_REQUESTED** [review]

**Comment by richardkchapman** [comment]
Are we going to merge this into 7.12.18 and tag rc2, or wait for 7.12.20 ?

File: rtl/eclrtl/rtldynfield.cpp (hunk 1 above)
Comment (jakesmith) [reply rtl/eclrtl/rtldynfield.cpp:51]: @richardkchapman - It's not quite ignoring, to ignore completely would mean standard (payload) translation was on,
it's treating it to mean turn translation off.

How about:
```
Unsupported alwaysECL translation mode used globally, translation disabled - use with HINT to set locally.
```

Comment (richardkchapman) [reply rtl/eclrtl/rtldynfield.cpp:51]: @jakesmith Works for me...

**Comment by jakesmith** [comment]
@richardkchapman - changed wording of warning, squashed in.

**Comment by richardkchapman** [comment]
@ghalliday Please sanity-check

**Review by richardkchapman — APPROVED** [review]

File: rtl/eclrtl/rtldynfield.cpp [hunk 2]
```diff
@@ -41,17 +41,16 @@ extern ECLRTL_API RecordTranslationMode getTranslationMode(const char *val, bool
     else if (strieq(val, "alwaysDisk") || strieq(val, "disk"))
     {
         if (!isLocal)
-            throw makeStringException(0, "alwaysDisk translation mode can only be set via HINT");
+            WARNLOG("alwaysDisk translation mode should only ever be used via a HINT");
         return RecordTranslationMode::AlwaysDisk;
     }
     else if (strieq(val, "alwaysECL") || strieq(val, "ecl"))
     {
         if (!isLocal)
```
Comment (ghalliday) [inline rtl/eclrtl/rtldynfield.cpp:49]: shouldn't this be
   if (isLocal)

**Review by ghalliday — COMMENTED** [review]
@jakesmith

File: rtl/eclrtl/rtldynfield.cpp (hunk 2 above)
Comment (jakesmith) [reply rtl/eclrtl/rtldynfield.cpp:49]: hm, thanks, yes it was in 1 iteration, thanks for spotting!

**Comment by jakesmith** [comment]
@ghalliday - address issue you spotted.

**Review by ghalliday — APPROVED** [review]
@richardkchapman looks good.

**Pull request merged**